diff --git a/cmake/SymEngineConfig.cmake.in b/cmake/SymEngineConfig.cmake.in
index dbfc80ba..d3a390b8 100644
--- a/cmake/SymEngineConfig.cmake.in
+++ b/cmake/SymEngineConfig.cmake.in
@@ -109,11 +109,11 @@ endif()

 list(REMOVE_DUPLICATES SYMENGINE_INCLUDE_DIRS)

-foreach(LIB "@SYMENGINE_TARGETS@")
-    # Remove linking of dependencies to later add them as targets
-    set_target_properties(${LIB} PROPERTIES IMPORTED_LINK_INTERFACE_LIBRARIES_RELEASE "")
-    set_target_properties(${LIB} PROPERTIES IMPORTED_LINK_INTERFACE_LIBRARIES_DEBUG "")
-endforeach()
+#foreach(LIB "@SYMENGINE_TARGETS@")
+#    set_target_properties(${LIB} PROPERTIES IMPORTED_LINK_INTERFACE_LIBRARIES_RELEASE "")
+#    set_target_properties(${LIB} PROPERTIES IMPORTED_LINK_INTERFACE_LIBRARIES_DEBUG "")
+#endforeach()
+set_target_properties(symengine PROPERTIES INTERFACE_LINK_LIBRARIES "${SYMENGINE_LIBRARIES}")

 set(SYMENGINE_LIBRARIES @SYMENGINE_TARGETS@ ${SYMENGINE_LIBRARIES})
 set(SYMENGINE_BUILD_TYPE "@CMAKE_BUILD_TYPE@")
